File: stack/nearest_smallest_to_right.py
def solve(arr):
    stack = []
    res = []

    for i in arr[::-1]:
        if not stack:
            res.append(-1)
        elif stack[-1] < i:
            res.append(stack[-1])
        else:
            while stack and stack[-1] >= i:
                stack.pop()
            res.append(-1) if not stack else res.append(stack[-1])
        stack.append(i)
    return res[::-1]


arr = [1, 3, 1, 4, 2, 1, 2, 4, 5]
res = solve(arr)

print(f'res -> {res}')

File: javascript/array/nearest_smallest_to_right.js
const arr = [2, 1, 0, 3, 4];

const solve = (arr) => {
  let [stack, res] = [[], []];
  const top = (stack) => stack[stack.length - 1];
  for (let i = arr.length - 1; i > -1; i--) {
    if (!stack[i]) res.push(-1);
    else if (top(stack) < arr[i]) {
      res.push(top(stack));
    } else if (!stack || top(stack) >= arr[i]) {
      while (!stack || top(stack) >= arr[i]) {
        stack.pop();
      }
      !stack ? res.push(-1) : res.push(top(stack));
    }
    stack.push(arr[i]);
  }
  return res.reverse();
};

const res = solve(arr);
console.log(`res`, res);
